File: app/src/main/java/com/noiseapps/itassistant/utils/Comparators.java
package com.noiseapps.itassistant.utils;

import java.util.Comparator;

import com.noiseapps.itassistant.model.jira.issues.Issue;

public class Comparators {

    public static class ISSUE {
        public static final Comparator<Issue> BY_KEY = (lhs, rhs) -> lhs.getKey().compareTo(rhs.getKey());
        public static final Comparator<Issue> BY_TYPE = (l, r) -> l.getFields().getIssueType().getId().compareTo(r.getFields().getIssueType().getId());
        public static final Comparator<Issue> BY_PRIORITY = (l, r) -> l.getFields().getPriority().getId().compareTo(r.getFields().getPriority().getId());
        public static final Comparator<Issue> BY_SUMMARY = (l, r) -> l.getFields().getSummary().compareTo(r.getFields().getSummary());
        public static final Comparator<Issue> BY_ASSIGNEE = ISSUE::compareByAssignee;
        public static final Comparator<Issue> BY_MODIFIED = (l, r) -> l.getFields().getUpdated().compareTo(r.getFields().getUpdated());
        public static final Comparator<Issue> BY_CREATED = (l, r) -> l.getFields().getCreated().compareTo(r.getFields().getCreated());


        private static int compareByAssignee(Issue l, Issue r) {
            String lAssignee = l.getFields().getAssignee() == null ? "" : l.getFields().getAssignee().getName();
            String rAssignee = r.getFields().getAssignee() == null ? "" : r.getFields().getAssignee().getName();
            return lAssignee.compareTo(rAssignee);
        }
    }
}

File: app/src/main/kotlin/com/noiseapps/itassistant/utils/Comparators.kt
package com.noiseapps.itassistant.utils

import com.noiseapps.itassistant.model.jira.issues.Issue
import java.util.*

public class Comparators {

    public object ISSUE {
        @JvmField public val BY_KEY: Comparator<Issue> = Comparator { l, r -> l.key.compareTo(r.key) }
        @JvmField public val BY_TYPE: Comparator<Issue> = Comparator { l, r -> l.fields.issueType.id.compareTo(r.fields.issueType.id) }
        @JvmField public val BY_PRIORITY: Comparator<Issue> = Comparator { l, r -> l.fields.priority.id.compareTo(r.fields.priority.id) }
        @JvmField public val BY_SUMMARY: Comparator<Issue> = Comparator { l, r -> l.fields.summary.compareTo(r.fields.summary) }
        @JvmField public val BY_ASSIGNEE: Comparator<Issue> = Comparator { l, r -> compareByAssignee(l, r) }
        @JvmField public val BY_MODIFIED: Comparator<Issue> = Comparator { l, r -> l.fields.updated.compareTo(r.fields.updated) }
        @JvmField public val BY_CREATED: Comparator<Issue> = Comparator { l, r -> l.fields.created.compareTo(r.fields.created) }

        private fun compareByAssignee(l: Issue, r: Issue): Int {
            val lAssignee = if (l.fields.assignee == null) "" else l.fields.assignee.name
            val rAssignee = if (r.fields.assignee == null) "" else r.fields.assignee.name
            return lAssignee.compareTo(rAssignee)
        }
    }


}
